### ISSUE TITLE
Keras API extensions - more flexible models

### ISSUE THREAD

**Issue by MillionIntegrals** [keras-team/keras#10039]
I'm a happy long-time user of keras and while many models can be expressed idiomatically in keras I see a certain inflexibility in the way keras handles **loss functions** and **metrics**. 

Ideally, I would like to see a more generic Model class, where **targets** (e.g. placeholders for ground truth data) are defined separately from **outputs** and user can specify any loss or metric as a function of **inputs** and **targets** and **outputs** just happen to be tensors that get returned from training.

Currently, we have a slightly more rigid structure where all **targets**, **lossess** and **metrics** are tightly coupled with outputs which is not a natural way to express many models.

I'm aware of `model.add_loss` method, this is what I'm currently using for implementing my models, but it seems that the additional loss values are not shown during training, and extra **targets** need to be specified as inputs to the model. 

I'd be really happy to implement such API for keras, but I wanted to gather your opinions first, if you think such extension would be a good part of the framework and how the best it would fit in the current design.

**Comment by fchollet** [comment]
What would your API look like?

> I'm aware of model.add_loss method, this is what I'm currently using for implementing my models, but it seems that the additional loss values are not shown during training

They're added to the main loss, which is displayed. Agreed that this may not be enough if you have multiple losses and you want to log them all.

> and extra targets need to be specified as inputs to the model.

Not necessarily, they could come from external data tensors as well (especially in TF).

**Comment by fchollet** [comment]
If you have complicated losses, you might be better off writing your own training loop (in which case you are free to use whatever losses and metrics you want and log them appropriately).

otherwise, I suggest decoupling inputs/outputs and loss computation, such as with the following template:

```python
input_data_1 = Input(shape_1)
[...]
input_data_n = Input(shape_n)

l1 = Conv2d(...)(input_data1)
[...]
output_layer_1 = Activation('softmax')(ln)
output_layer_k = Activation('softmax')(ln)

target_1 = Input(target_shape_1)  # You can use `K.placeholder()` too
[...]
target_j = Input(target_shape_j)

# Outputs are what you want to retrieve from your model when running `predict`
model = Model(
  inputs = [input_1, ..., input_n],
  outputs = [output_layer_1, ..., output_layer_k])

model.add_loss(binary_crossentropy(output_layer_1, target_1))
model.add_loss(custom_loss(output_layer_2, target_2, target_3))
# etc.

model.compile(optimizer)  # no external loss

# Only for TF: pass targets via `feed_dict`
feed_dict = {target_1: target_1_value, ...}
model.train_on_batch([input_value_1, ..., input_value_n], session_kwargs={'feed_dict': feed_dict})
```

However this does not cover metrics, or logging individual losses. We could rather easily add support for logging losses added with `add_loss` though, and we could add a `add_metric` method for arbitrary metrics.

Alternatively, if you are using TF data tensors, you can do the following (in `tf.keras` at HEAD):

```python
model_input = Input(shape_1)
...
model_output = ...
model = Model(model_input, model_output)

# Using the tf.data API
inputs, targets = dataset.get_next()
model.add_loss(loss_fn(targets, model_output))

model.compile(optimizer)
model.fit(inputs, epochs=10, steps_per_epoch=1000)
```

This is pretty clean and intuitive. Still doesn't support logging of individual custom losses nor custom metrics.

**Comment by MillionIntegrals** [comment]
Thank you for all the advice, they all seem like valid solutions to the majority of the problem, but feel a bit like workarounds rather than idiomatic solution. There seems to be a bit of tinkering always involved to get exactly what I need.

I don't really know how common it is in the community, but it's a pretty common use case for me.
Therefore, just like you've mentioned, I'd be happy to implement my own training loop wrapped in a slightly modified Model class so that I can have a setup I need for my experiments working and clean.

My main question is if there are more people which would like to use the more generic API and if you think there is a place for it in `keras` or in `keras-contrib` or should I just make my own, separate package for that?

**Comment by dumkar** [comment]
I have exactly the same issue, and also see some others users asking for this (especially since VAEs are getting more popular). See e.g. ([here](https://stackoverflow.com/questions/54069363/output-multiple-losses-added-by-add-loss-in-keras/54109533#54109533))
